### PR TITLE
fix(git): keep every commit in verbose git log instead of truncating (#199)

### LINKED ISSUE #199: git log (verbose, multi-commit) drops 10 of 12 commits and cuts mid-line — no count, inflates the git savings figure
## Summary

A verbose `git log` (default format, not `--oneline`) with N commits is distilled down to the first ~2 commits' full bodies and the rest are dropped, with the output cut mid-line and no count. `git log --oneline` is fine (#192 keeps subjects); this is the multi-commit verbose path, a different code path that truncates.

## Reproduction (current main, 12b1d28)

```
$ git log -12 | grep -c '^commit '
12
$ git log -12 | OMNI_CMD="git log -12" ./target/debug/omni 2>/dev/null | grep -c '^commit '
2
```

345 lines / 17,703 B → 19 lines / 1,361 B, reported as **92% saved** — but 10 of the 12 commits are gone, and the surviving output ends mid-sentence (`… TestDistiller's headline fell`), with no `N more commits` marker.

## Why it matters

- **Lost signal on a command people run constantly.** The dropped commits are the answer.
- **False completeness** (#111 class): no count, so a reader cannot tell 10 commits were dropped.
- **It inflates the published git figure.** The benchmark shows `git 91%`; part of that "saving" is dropped commits, not removed noise — relevant to #184 (honest numbers). A re-benchmark should treat this as lost signal, not compression.

## Contrast: the lossless paths work

- `git log --oneline -20` → 20 → 20 lines, all subjects kept (0% "saved", correct).
- `git diff HEAD~4` → 9% (mostly passthrough, lossless).

So the defect is specific to verbose multi-commit `git log`.

## Suspected cause

Same shape as #198: a large multi-line input that does not match the oneline/`git show` paths lands in a line/byte-cap fallback that truncates without a count. Whichever cap fires must state `N of M commits shown` and route the tail to the RewindStore (#111 invariant), or the verbose log must keep one compact line per commit (hash + subject) the way `--oneline` does.

## Environment

- `omni` current main (12b1d28), debug build, macOS. Found while building an honest git demo GIF; this is why the git demo was not shipped.

### PR DESCRIPTION
**Pull request by fajarhide**
Closes #199 · completes the enumeration/lost-signal set with #198 (#203, merged) under #200

## What
`git log -12` returned **2 of 12 commits**, cut mid-line, no marker. The #110 shadow trap: `signals/tools/git_log.toml` matched `^git log` and won ahead of the Rust distiller, and its line-strip rules cannot tell a commit subject from a body line (both 4-space indented), so on a verbose log it stripped only `Author:`/`Date:`, kept every body, blew past `max_lines = 20`, and truncated the older commits away.

Remove the filter. Verbose `git log` now falls through to `distill_log`, which keeps **one compact `hash subject` line per commit** — the `--oneline` view — so every commit survives while body + `Author`/`Date`/`Merge` go. `--oneline` was already handled by the Rust distiller (the filter matched nothing there, #107), so it's unaffected.

`distill_log` also now **fails open**: the fixture harness feeds arbitrary text through it as `git status`, and the old "first 5 lines" last resort became returning the input verbatim (#143), so non-log input is never reduced to empty or a lossy guess.

## Verified
Through the binary: `git log -12` → 12/12 commits (345 lines → 11), `git log --oneline -8` → all 8. Regression test asserts a three-commit log with bodies keeps all subjects and drops bodies. The removed filter's integration tests went with it. Local (Homebrew rust 1.94.0, not CI's 1.97.0): fmt, clippy `--all-targets -D warnings`, full suite (1,150 passed), `tests/smoke_test.sh` (31/31).

With #203 (merged), this closes the enumeration/lost-signal work under #200: `find`/`ls`/`wc`/`env`/`df` pass through, verbose `git log` keeps every commit. Remaining under #200: the `ps aux` `MAX_OUTPUT_BYTES` cap (#111-class), tracked separately.

## Summary
Fixed 4 bugs in OMNI's output distillers: removed a broken TOML filter that dropped commits from verbose `git log`, made enumeration commands (`ls`, `find`, `wc`, `env`, `df`) pass through verbatim, implemented the `priority` field that was parsed but never used, and prevented `TestDistiller` from fabricating "0 passed, 0 failed" for non-test output.

---

## Key Changes

| Component | Change |
|-----------|--------|
| `git_log.toml` | **DELETED** — its line-strip rules couldn't distinguish subjects from body lines, causing commit loss |
| `distill_log` (Rust) | Now keeps one `hash subject` line per commit; drops bodies and metadata; fails open on non-log input |
| `distill_with_command` | Enumeration commands (`ls`, `find`, `ps`, `wc`, `env`, `df`, `du`, `tree`, `stat`) now pass through verbatim |
| Filter loading | `priority` now actually sorts filters; ties broken deterministically (alphabetical) instead of hash-random |
| `TestDistiller` | Returns input verbatim instead of fabricating "0 passed, 0 failed" when no tally is parsed |

---

## Detailed Breakdown

**`git log` fix (#199)**
- Removed `signals/tools/git_log.toml` (matched `^git log`, stripped `Author:`/`Date:`/`Merge:`, kept body lines, hit `max_lines = 20`, truncated older commits)
- `distill_log` now extracts one compact `hash subject` line per commit (mirrors `--oneline` behavior)
- Commits are no longer lost: `git log -12` returns all 12 (was 2)
- Falls back to input verbatim on unrecognized input instead of returning empty or lossy guess

**Enumeration commands pass-through (#198, #200)**
- `ls -la`, `find`, `wc`, `env`, `df`, `du`, `tree`, `ps`, `stat` now route to verbatim passthrough
- Collapse hooks (`hooks/pipe.rs`, `hooks/post_tool.rs`) skipped via `is_enumeration_command` to prevent re-compression into `[N similar lines collapsed]`
- `grep`/`rg` intentionally excluded (their distiller hoists path losslessly)
- Residual: very large output (`ps aux`, ~148 KB) still capped by `MAX_OUTPUT_BYTES`

**Filter `priority` implemented (#119, #134)**
- `load_all_filters` now sorts by descending `priority` with stable sort
- Filter map changed from `HashMap` to `BTreeMap` for deterministic load order
- Equal priorities break ties alphabetically instead of arbitrarily
- Test updated to use non-priority-sorted load order to catch broken sorts

**TestDistiller fabrication fix (#195)**
- When `summary.is_none()` and nothing counted, returns input verbatim instead of `"Tests: 0 passed, 0 failed"`
- Real zero-test runs (e.g., `cargo test` in empty crate) remain unaffected — they have a `summary` line
- Reproduction: `go test ./...` across packages with no test files now passes through `[no test files]` instead of fabricating a tally

---

## Notes
- Snapshot tests for `ls`/`find`/`env` were replaced with explicit passthrough assertions
- Removed filter's integration tests (`toml_filter_integration.rs`) and entry from `expected` list
- Full suite: 1,150 tests pass on Rust 1.94.0+

---

## Breaking Changes
- **`git log` output format changed**: verbose logs now show `hash subject` per commit (like `--oneline`) instead of full commit blocks with bodies. This is intentional — previous behavior dropped commits entirely.
- **TOML filter `git_log` removed**: configs referencing it will no longer match; OMNI falls back to Rust distiller automatically.

_Last updated: 2026-07-26 03:59:09_

<details open><summary><h3>Greptile Summary</h3></summary>

Removes the `signals/tools/git_log.toml` TOML filter that was silently truncating verbose `git log` output (the filter's `max_lines = 20` was overflowed by body lines it couldn't distinguish from subjects), and replaces it with a proper state machine in `distill_log` that emits exactly one `hash subject` line per commit. The fail-open path is also improved from a "first 5 lines" guess to returning the input verbatim when no commits are recognised.

- **TOML filter deleted**: `git_log.toml` and its integration tests are removed; verbose `git log` now routes entirely through the Rust `distill_log`, which was already the sole handler for `--oneline` format.
- **State machine added**: the `awaiting_subject` flag captures the first non-metadata line after each `commit <sha>` header as the subject, then drops all remaining body lines — keeping every commit while cutting the volume that previously overwhelmed `max_lines`.
- **Regression test added**: `verbose_log_keeps_every_commit_and_drops_the_body` asserts a three-commit log with multi-line bodies produces exactly 3 output lines, all subjects intact, no body text.
</details>

<details open><summary><h3>Confidence Score: 4/5</h3></summary>

Safe to merge; the core fix is correct and well-tested. One narrow edge case in the state machine is worth addressing before widespread use.

The state machine correctly keeps every commit's subject and drops bodies in all normal inputs. The one gap is that the RE_GIT_LOG_HASH branch does not reset awaiting_subject, so a verbose commit whose subject line starts with 7+ lowercase hex characters followed by a space will leave the flag set, causing the first body line to also appear in the output. This is a rare input shape and produces no data loss, but it is a present defect in the changed logic path.

**Files Needing Attention:** src/distillers/git.rs — the RE_GIT_LOG_HASH branch in distill_log
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/distillers/git.rs | Rewrites distill_log with an awaiting_subject state machine that captures exactly one subject line per commit; adds fail-open fallback and a regression test. One edge case: the RE_GIT_LOG_HASH branch doesn't reset awaiting_subject, so a verbose subject matching the oneline pattern can leak the next body line. |
| signals/tools/git_log.toml | Deleted: the faulty TOML filter whose line-strip rules could not distinguish commit subjects from body lines, causing verbose git log to blow past max_lines=20 and drop older commits silently. |
| tests/toml_filter_integration.rs | Removes the git_log filter integration test module and its entry in the all_filter_files_loaded list, correctly aligned with the deleted TOML file. |
| CHANGELOG.md | Adds a detailed entry under [Unreleased] Fixed explaining root cause, fix, measurement, and verification for #199. |

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[git log output] --> B{TOML filter match?}
    B -- "before PR: git_log.toml wins" --> C[strip Author/Date/Merge\nkeep all other lines\nmax_lines=20]
    C --> D["❌ body lines kept → overflow\n→ older commits truncated"]
    B -- "after PR: no filter" --> E[distill_log]
    E --> F{line type?}
    F -- "commit sha..." --> G[emit 7-char hash + space\nawaiting_subject = true]
    F -- "Author/Date/Merge" --> H[drop]
    F -- "oneline RE match" --> I[emit full line]
    F -- "awaiting_subject = true" --> J[emit as subject\nawaiting_subject = false]
    F -- "other body line" --> K[drop]
    G --> F
    J --> F
    E --> L{any commits found?}
    L -- yes --> M["compact hash subject\none line per commit ✅"]
    L -- no --> N[return input verbatim\nfail-open #143]
```
</details>

<a href="https://app.greptile.com/ide/claude-code?prompt=Fix%20the%20following%201%20code%20review%20issue.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%201%0Asrc%2Fdistillers%2Fgit.rs%3A187-207%0A**%60awaiting_subject%60%20not%20cleared%20in%20%60RE_GIT_LOG_HASH%60%20branch**%0A%0AWhen%20a%20verbose%20log%20commit's%20subject%20line%20happens%20to%20match%20the%20oneline%20regex%20%28e.g.%2C%20%60a1b2c3d%20add%20migration%20for%20...%60%20%E2%80%94%207%2B%20lowercase%20hex%20chars%20then%20a%20space%29%2C%20the%20%60RE_GIT_LOG_HASH%60%20branch%20fires%20and%20appends%20the%20line%20correctly%2C%20but%20%60awaiting_subject%60%20is%20never%20set%20to%20%60false%60.%20On%20the%20very%20next%20non-metadata%2C%20non-empty%20line%20%28the%20commit%20body%29%2C%20the%20%60else%20if%20awaiting_subject%60%20branch%20then%20also%20fires%2C%20appending%20that%20first%20body%20line%20as%20if%20it%20were%20the%20subject.%20Adding%20%60awaiting_subject%20%3D%20false%3B%60%20inside%20the%20RE%20branch%20prevents%20the%20leak.%20In%20%60--oneline%60%20format%20%60awaiting_subject%60%20is%20always%20%60false%60%20when%20this%20branch%20runs%2C%20so%20the%20change%20is%20safe%20for%20both%20formats.%0A%0A&repo=fajarhide%2Fomni&pr=204&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=6">Fix All in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22fajarhide%2Fomni%22%20on%20the%20existing%20branch%20%22fix%2F199-git-log-verbose%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22fix%2F199-git-log-verbose%22.%0A%0AFix%20the%20following%201%20code%20review%20issue.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%201%0Asrc%2Fdistillers%2Fgit.rs%3A187-207%0A**%60awaiting_subject%60%20not%20cleared%20in%20%60RE_GIT_LOG_HASH%60%20branch**%0A%0AWhen%20a%20verbose%20log%20commit's%20subject%20line%20happens%20to%20match%20the%20oneline%20regex%20%28e.g.%2C%20%60a1b2c3d%20add%20migration%20for%20...%60%20%E2%80%94%207%2B%20lowercase%20hex%20chars%20then%20a%20space%29%2C%20the%20%60RE_GIT_LOG_HASH%60%20branch%20fires%20and%20appends%20the%20line%20correctly%2C%20but%20%60awaiting_subject%60%20is%20never%20set%20to%20%60false%60.%20On%20the%20very%20next%20non-metadata%2C%20non-empty%20line%20%28the%20commit%20body%29%2C%20the%20%60else%20if%20awaiting_subject%60%20branch%20then%20also%20fires%2C%20appending%20that%20first%20body%20line%20as%20if%20it%20were%20the%20subject.%20Adding%20%60awaiting_subject%20%3D%20false%3B%60%20inside%20the%20RE%20branch%20prevents%20the%20leak.%20In%20%60--oneline%60%20format%20%60awaiting_subject%60%20is%20always%20%60false%60%20when%20this%20branch%20runs%2C%20so%20the%20change%20is%20safe%20for%20both%20formats.%0A%0A&repo=fajarhide%2Fomni&pr=204&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodexDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodex.svg?v=6">Fix All in Codex</picture></a> <a href="https://app.greptile.com/api/ide/cursor?prompt=Fix%20the%20following%201%20code%20review%20issue.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%201%0Asrc%2Fdistillers%2Fgit.rs%3A187-207%0A**%60awaiting_subject%60%20not%20cleared%20in%20%60RE_GIT_LOG_HASH%60%20branch**%0A%0AWhen%20a%20verbose%20log%20commit's%20subject%20line%20happens%20to%20match%20the%20oneline%20regex%20%28e.g.%2C%20%60a1b2c3d%20add%20migration%20for%20...%60%20%E2%80%94%207%2B%20lowercase%20hex%20chars%20then%20a%20space%29%2C%20the%20%60RE_GIT_LOG_HASH%60%20branch%20fires%20and%20appends%20the%20line%20correctly%2C%20but%20%60awaiting_subject%60%20is%20never%20set%20to%20%60false%60.%20On%20the%20very%20next%20non-metadata%2C%20non-empty%20line%20%28the%20commit%20body%29%2C%20the%20%60else%20if%20awaiting_subject%60%20branch%20then%20also%20fires%2C%20appending%20that%20first%20body%20line%20as%20if%20it%20were%20the%20subject.%20Adding%20%60awaiting_subject%20%3D%20false%3B%60%20inside%20the%20RE%20branch%20prevents%20the%20leak.%20In%20%60--oneline%60%20format%20%60awaiting_subject%60%20is%20always%20%60false%60%20when%20this%20branch%20runs%2C%20so%20the%20change%20is%20safe%20for%20both%20formats.%0A%0A&pr=204&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursorDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursor.svg?v=6">Fix All in Cursor</picture></a> <a href="https://app.greptile.com/api/ide/conductor?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22fajarhide%2Fomni%22%20on%20the%20existing%20branch%20%22fix%2F199-git-log-verbose%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22fix%2F199-git-log-verbose%22.%0A%0AFix%20the%20following%201%20code%20review%20issue.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%201%0Asrc%2Fdistillers%2Fgit.rs%3A187-207%0A**%60awaiting_subject%60%20not%20cleared%20in%20%60RE_GIT_LOG_HASH%60%20branch**%0A%0AWhen%20a%20verbose%20log%20commit's%20subject%20line%20happens%20to%20match%20the%20oneline%20regex%20%28e.g.%2C%20%60a1b2c3d%20add%20migration%20for%20...%60%20%E2%80%94%207%2B%20lowercase%20hex%20chars%20then%20a%20space%29%2C%20the%20%60RE_GIT_LOG_HASH%60%20branch%20fires%20and%20appends%20the%20line%20correctly%2C%20but%20%60awaiting_subject%60%20is%20never%20set%20to%20%60false%60.%20On%20the%20very%20next%20non-metadata%2C%20non-empty%20line%20%28the%20commit%20body%29%2C%20the%20%60else%20if%20awaiting_subject%60%20branch%20then%20also%20fires%2C%20appending%20that%20first%20body%20line%20as%20if%20it%20were%20the%20subject.%20Adding%20%60awaiting_subject%20%3D%20false%3B%60%20inside%20the%20RE%20branch%20prevents%20the%20leak.%20In%20%60--oneline%60%20format%20%60awaiting_subject%60%20is%20always%20%60false%60%20when%20this%20branch%20runs%2C%20so%20the%20change%20is%20safe%20for%20both%20formats.%0A%0A&repo=fajarhide%2Fomni"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInConductorDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInConductor.svg?v=6">Fix All in Conductor</picture></a>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 1 code review issue. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 1
src/distillers/git.rs:187-207
**`awaiting_subject` not cleared in `RE_GIT_LOG_HASH` branch**

When a verbose log commit's subject line happens to match the oneline regex (e.g., `a1b2c3d add migration for ...` — 7+ lowercase hex chars then a space), the `RE_GIT_LOG_HASH` branch fires and appends the line correctly, but `awaiting_subject` is never set to `false`. On the very next non-metadata, non-empty line (the commit body), the `else if awaiting_subject` branch then also fires, appending that first body line as if it were the subject. Adding `awaiting_subject = false;` inside the RE branch prevents the leak. In `--oneline` format `awaiting_subject` is always `false` when this branch runs, so the change is safe for both formats.

`````

</details>

<sub>Reviews (1): Last reviewed commit: ["Merge branch &#39;main&#39; into fix/199-git-log..."](https://github.com/fajarhide/omni/commit/843ab11308a6e31a2378f9b9ee386c205172e3b0) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=47318952)</sub>

> Greptile also left **1 inline comment** on this PR.

**Context used:**

- Context used - CLAUDE.md ([source](https://app.greptile.com/weekndlabs/github/fajarhide/omni/-/custom-context?memory=1653942e-45b5-4251-8409-19e160d03e6e))
- Context used - AGENTS.md ([source](https://app.greptile.com/weekndlabs/github/fajarhide/omni/-/custom-context?memory=5b7b5156-798e-47d0-935c-5cc5cabdcec2))

### CODE REVIEW

File: src/distillers/git.rs [hunk 1]
```diff
@@ -169,37 +176,45 @@ fn distill_log(segments: &[OutputSegment], _input: &str) -> String {
...
+                || line.starts_with("Merge:")
             {
+                // Metadata — drop.
+            } else if awaiting_subject {
+                // First content line after `commit <sha>` is the subject.
                 out.push_str(line);
                 out.push('\n');
+                awaiting_subject = false;
             }
+            // Any further body lines before the next `commit` are dropped (#199).
```
[the remark this answers is not in the document]
Comment (fajarhide) [reply src/distillers/git.rs:207]: Real leak. Fixed in #205 with a regression test proven to fail without the change.
